audition/distance_from_best.py

```diff
@@ -0,0 +1,300 @@
+    def __init__(self, distance_from_best_table):
+        """Generate a histogram illustrating the effect of different below-best maximum
+        thresholds across the dataset.
+
+        Args:
+            distance_from_best_table (audition.DistanceFromBestTable)
+                A pre-populated distance-from-best database table
+        """
+        self.distance_from_best_table = distance_from_best_table
+
+    def generate_histogram_data(
+        self,
+        metric,
+        metric_param,
+        model_group_ids,
+        train_end_times,
+    ):
+        """Fetch data necessary for producing the histogram from the distance table
+
+        Arguments:
+            metric (string) -- model evaluation metric, such as 'precision@'
+            metric_param (string) -- model evaluation metric parameter,
+                such as '300_abs'
+            model_group_ids (list) - Model group ids to include in the dataset
+            train_end_times (list) - Train end times to include in the dataset
+
+        Returns: (pandas.DataFrame) The relevant models and the percentage of time
+            each was within various thresholds of the best model at that time
+        """
+        model_group_union_sql = ' union all '.join([
+            '(select {} as model_group_id)'.format(model_group_id)
+            for model_group_id in model_group_ids
+        ])
+        sel_params = {
+            'metric': metric,
+            'metric_param': metric_param,
+            'model_group_union_sql': model_group_union_sql,
+            'distance_table': self.distance_from_best_table.distance_table,
+            'model_group_str': str_in_sql(model_group_ids),
+            'train_end_str': str_in_sql(train_end_times),
+        }
+        sel = """
+                with model_group_ids as ({model_group_union_sql}),
+                x_vals AS (
+                  SELECT m.model_group_id, s.pct_diff
+                  FROM
+                  (
+                  SELECT GENERATE_SERIES(0,100) / 100.0 AS pct_diff
+                  ) s
+                  CROSS JOIN
+                  (
+                  SELECT DISTINCT model_group_id FROM model_group_ids
+                  ) m
+                )
+                SELECT dist.model_group_id, pct_diff, mg.model_type,
+                       COUNT(*) AS num_models,
+                       AVG(CASE WHEN below_best <= pct_diff THEN 1 ELSE 0 END) AS pct_of_time
+                FROM {distance_table} dist
+                JOIN x_vals USING(model_group_id)
+                JOIN results.model_groups mg using (model_group_id)
+                WHERE
+                    dist.metric='{metric}'
+                    AND dist.parameter='{metric_param}'
+                    and model_group_id in ({model_group_str})
+                    and train_end_time in ({train_end_str})
+                GROUP BY 1,2,3
+            """.format(**sel_params)
+
+        return pd.read_sql(sel, self.distance_from_best_table.db_engine)
```

Comment:
here I would probably add a `.sort_values(['model_group_id', 'pct_diff'])` before returning the dataframe (or even just do the equivalent in the SQL above) just to ensure the plots render correctly.